### ISSUE TITLE
Show extra balloon notifications

### ISSUE THREAD

**Issue by canton7** [canton7/SyncTrayzor#18]
Discussed in https://discourse.syncthing.net/t/synctrayzor-windows-host-for-syncthing-installer-auto-start-built-in-browser-tray-icon-folder-watcher-and-more/2011/48.

Asker wants to see:
 - Device Connected
 - Device Disconnected
 - Sync notifications

We can easily do connected / disconnected and file transferred from remote -> local (DeviceConnectedEvent, DeviceDisconnectedEvent, ItemFinishedEvent), but we can't easily to local -> remote (we get a LocalIndexUpdated, followed by a RemoteIndexUpdated. But RemoteIndexUpdated just says how many files have changed - not which ones. It's also fired when a file is spontaneously changed on the remote side).

We also don't want to risk trying to pop up balloon messages too frequently - it gets annoying. Probably a custom list of events which automatically appends / etc. Our tray icon library allows this, but it will take a little bit of trial and error to make it look right.

**Comment by kreischweide** [comment]
Maybe have a queue for the messages and prioritize the different types?

- Important ones will be displayed immediately (clicking on them might pop a dialog with a question like accept/ignore/later)
- Lesser important ones will be grouped by event (like there were 5 completes updates to files in the last 3 minutes) and only spawned every X(configurable) minutes

You might as well turn of lesser importatnt ones in the settings.

**Comment by canton7** [comment]
I don't think we'll want to display such wordy things that they'll need to be accepted. These are just transient status updates. However grouping / categorising / ordering is probably a sensible thing to be doing!

I'm planning to have all of these customizable in the settings - even the Synced events annoy me :)

**Comment by canton7** [comment]
I'm planning on displaying device connected / disconnected in the same style as the current 'synced' notifications. For the 'file synced', I'm going to take a long hard look at what Dropbox does - we really run the risk of seriously pissing people when they sync lots of small files if we follow the current route for this.

**Comment by kreischweide** [comment]
Sure. Dropbox collects added files in single lists. had the chance to see it in action as we migrated some 500GB backups over there. On the origin you add 100k files, Dropbox uploads them to their system and as soon as a single one of them is uploaded, it triggers the streaming sync feature and streams the parts down to subscribed clients. The notification will be shown as soon as the whole file list is downloaded (100k files). Also on the website they group them in this lists.

Not sure if Syncthing has the same understanding of it, but I would assume it cares about the blocks of single files, so you would need to come up with some grouping mechanism.

The user experience should be as simple as possible. If I copy over a folder with alot of files, it should be only the overall status that I get notified about.

**Comment by canton7** [comment]
Right, the connectivity alerts are in v1.0.9. Now for developing file lists, where possible...

**Comment by canton7** [comment]
After a chat on IRC, it looks like we can tell when a file's been synchronized from a remote device to the local folder, but not the other way around. A remote device does not tell us when it's finished updating one of its files, which is essentially what we're asking for. It also wouldn't give any indication of *which* version is updated it file to - there's no guarentee that it picked up our change, rather than anyone else's.

@Rewt0r, is this still of interest, if we can only display 'downloaded' files?

**Comment by canton7** [comment]
FYI, this is next on my list, after #37.

**Comment by fastfwd** [comment]
Antony, SyncTrayzor is working great in my system.  Thank you for writing it!

I've sen an issue with the balloon notifications: If a laptop goes to sleep overnight, it looks as though its notifications are queued.  And then when the machine is awakened, all the queued "Device Connected/Disconnected" and "Finished Syncing" notifications slowly appear, one after the other.  They appear at least ten seconds apart, and my laptop typically wakes up with hundreds of waiting notifications, so it can take an hour for the notification queue to empty.

Maybe the queue can be pruned of duplicates, or of events older than a certain threshold or something?

**Comment by canton7** [comment]
If they're appearing 10 seconds apart, it looks like Syncthing's actually telling us 10 seconds apart - when they're properly stacking up, they appear a couple of seconds after one another. 

To be honest, I find the notifications largely annoying: that's why there's an option in the settings to turn them off ;) If you look at the log file (`C:\Users\<You>\AppData\Roaming\SyncTrayzor\logs\SyncTrayzor.log` for the installed version, or `data\logs` for the portable version) you can see Syncthing's events as they come in (prefixed with 'SyncThingEventWatcher') - you should see FolderSyncStateChanged, DeviceConnected, and DeviceDisconnected events come in. 

However, an hour *does* sound a bit extreme! 

Can you open a new issue, and give a excerpt from your log over a period when these balloons are are appearing once every 10 secs? That will nail down for sure whether Syncthing's delivering them periodically, or all at once.

**Comment by canton7** [comment]
Actually I know what this is. Sit tight until the next release, then let me know if it's still an issue.

**Comment by Rewt0r** [comment]
Hi @canton7, sorry to bump an old issue but I've just updated our whole cluster to the new Syncthing using SyncTrayzor over Syncthing-Tray and have noticed that the notifications for syncing doesn't display what file was synced. Could you incorporate this if it is not already in your planning? I like the idea of displaying something similar to DropBox/Google Drive.

Thanks for your work.

**Comment by canton7** [comment]
Yeah, I'm afraid this feature request is both a) a lot of work, and b) not actively breaking anything, which means it's continuously near the bottom of the priority list.

We can get two bits of information from Syncthing:

 1) We know when a folder is in the 'Syncing' state (as opposed to 'Scanning' or 'Idle')
 2) We get notifications when files are *downloaded* to the current machine

We currently use 1): a transition from Syncting -> anything else is regarded as "folder has finished synchronizing". This feature would need 2), and we'd only be able to notify the user about *downloaded* files, never *uploaded* ones (there are a few discussions about this in the Syncthing issues and on the forums).

**Comment by canton7** [comment]
Just an update: currently in the middle of this. Still got maybe a week's work left on it, here's a sneak peek:
![capture](https://cloud.githubusercontent.com/assets/568104/7766753/66dca03a-0064-11e5-8720-c98c3de2f2a1.PNG)

**Comment by AudriusButkevicius** [comment]
Superb!

**Comment by Rewt0r** [comment]
Looking good!

**Comment by canton7** [comment]
Righty, this will be in the next release. Let me know what you think of it!
